### PR TITLE
drivers: dai: Intel dmic driver

### PR DESCRIPTION
**Pull request by abonislawski**
Intel DMIC driver originally submitted here:
https://github.com/zephyrproject-rtos/zephyr/pull/45563

This PR contains the same driver + volume rampup and unmute delay
and DMIC sync period review changes
FYI @lgirdwood @kv2019i @tmleman @GeorgeCGV @teburd @nashif

### CODE REVIEW

**Review by tmleman — DISMISSED** [review]

File: drivers/dai/intel/dmic/Kconfig.dmic [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# SOF SSP configuration options
```
Comment (nashif) [inline drivers/dai/intel/dmic/Kconfig.dmic:1]: please use same structure used in other drivers, here I would put this file in drivers/dai/dmic/Kconfig.dmic_intel 
and the source/header files in drivers/dai/dmic/dmic_intel.c and drivers/dai/dmic/dmic_intel.h.

**Review by nashif — DISMISSED** [review]

File: drivers/dai/intel/dmic/Kconfig.dmic (hunk 1 above)
Comment (abonislawski) [reply drivers/dai/intel/dmic/Kconfig.dmic]: Are you sure about dmic_intel?
In dai we have
alh - alh.c, Kconfig.alh
ssp - ssp.c Kconfig.ssp
dmic - dmic.c Kconfig.dmic

So only DMIC will be different with _intel

Changed "SOF SSP configuration" to "SOF DMIC configuration"

Comment (nashif) [reply drivers/dai/intel/dmic/Kconfig.dmic]: I am looking now at drivers/dai and see the intel folder already exists in there, so this is following the same model. making  dmic different is weird now, we can move things around later if needed. good as for now.

**Review by tmleman — DISMISSED** [review]

File: include/zephyr/drivers/dai.h [hunk 2]
```diff
@@ -160,6 +160,11 @@ enum dai_trigger_cmd {
 	 * state.
 	 */
 	DAI_TRIGGER_RESET,
+	/** @brief Copy
+	 *
+	 * This trigger prepares for data copying.
+	 */
+	DAI_TRIGGER_COPY,
```
[the remark this answers is not in the document]
Comment (abonislawski) [reply include/zephyr/drivers/dai.h:167]: yeap, it should be _COPY in gain ramp, fixed

File: drivers/dai/intel/dmic/dmic.h [hunk 3]
```diff
@@ -0,0 +1,648 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __INTEL_DAI_DRIVER_DMIC_H__
+#define __INTEL_DAI_DRIVER_DMIC_H__
+
+#include <zephyr/sys/util_macro.h>
+
+/* initialization priority */
+#define DMIC_DAI_INIT_PRIORITY 32
```
[the remark this answers is not in the document]
Comment (abonislawski) [reply drivers/dai/intel/dmic/dmic.h]: Fixed, actually there is a general CONFIG_DAI_INIT_PRIORITY in dai Kconfig so just used it

File: drivers/dai/intel/dmic/dmic.h [hunk 4]
```diff
@@ -0,0 +1,648 @@
...
+
+#define SET_BIT(b, x)		(((x) & 1) << (b))
+
+#define SET_BITS(b_hi, b_lo, x)	\
+	(((x) & ((1ULL << ((b_hi) - (b_lo) + 1ULL)) - 1ULL)) << (b_lo))
+
+#define GET_BIT(b, x) \
+	(((x) & (1ULL << (b))) >> (b))
+
+#define GET_BITS(b_hi, b_lo, x) \
```
[the remark this answers is not in the document]
Comment (teburd) [reply drivers/dai/intel/dmic/dmic.h:24]: Yes of course

**Comment by abonislawski** [comment]
Nothing changed, just rebase to fix unrelated CI errors

**Review by teburd — APPROVED** [review]
this looks good to me, I'm trusting its been tested in some way

**Review by tmleman — APPROVED** [review]

**Pull request merged**